### ISSUE TITLE
Static site sub-routes missing interactivity/not playing well with hydration

### ISSUE THREAD

**Issue by Tehnix** [leptos-rs/leptos#2663]
**Describe the bug**
I've been exploring SSG with Leptos a bit, trying to document my steps along the way. I've got everything _mostly_ working, except for reloading a statically generated page.

One thing that's broken though seems to be interactivity on pages when you navigate directly to them or reload while on them.

My suspicion is that it has something to do with hydration and a mismatch between the element IDs that it's looking for to reestablish interactivity. Not too sure though 😅 

**Leptos Dependencies**

I've done some slight adjustments to the standard [start-axum](https://github.com/leptos-rs/start-axum) template, which you can find here https://github.com/Tehnix/comparison-ssg-rust-js/tree/main/leptos-example.

**To Reproduce**

1. Clone the example repo `git clone git@github.com:Tehnix/comparison-ssg-rust-js.git`
2. Go into the Leptos project `cd leptos-example`
3. Build the Leptos project `cargo leptos build --release`
4. Run the binary to generate the static routes `./target/release/leptos-example`
5. You should now have files in `./target/site`
6. Serve the files via e.g. `(cd target/site && bunx simplehttpserver)`
7. Visit a sub-page directly in the browser http://localhost:8000/another-page.html
8. Interactivity is broken because hydration can't find the elements/components

**Expected behavior**

I would expect sub-routes to hydrate correctly and support interactivity, such as using signals etc.

When navigating to the root route at http://localhost:8000 everything works, and the sub-routes also work when navigating to them directly from the links on the root page. Things break when we reload on any of the sub-routes where hydration seems to fail (at least that's my guess as to what causes the lack of interactivity).

**Screenshots**

Loading a sub-route directly, e.g. http://localhost:8000/another-page.html, failing hydration and not calling the WASM/JS:

<img width="953" alt="Screenshot 2024-07-02 at 16 54 37" src="https://github.com/leptos-rs/leptos/assets/1189998/d8000cb8-012c-4097-813a-8ce4a58d72d0">

Loading a sub-route by loading the root page at http://localhost:8000 and then navigating to it, with everything working (note the `328` indicating it can fetch the `window.innerHeight` value):

<img width="953" alt="Screenshot 2024-07-02 at 16 57 11" src="https://github.com/leptos-rs/leptos/assets/1189998/9e6ec19c-7ab1-4ae3-ad63-b4d2da1cc88e">

**Additional context**

**Comment by Tehnix** [comment]
Ah, that sorta makes sense 😅 (at least for why hydration isn't kicking in and acting like it's looking for the elements on the root page)

My goal is to build static files and deploy/host it on e.g. AWS S3 or CloudFlare Pages, so it's necessary for me to not have any binary serve it, but indeed get it working with any random thing that can serve HTML.

I've added `trailing_slash=TrailingSlash::Exact` do my `Router` so that it would expect the path to end on `/` (e.g. `http://127.0.0.1:8000/another-page/`) since that's what will happen when moving the files to be at `target/site/another-page/index.html`:

```
        <Router fallback=|| {
            let mut outside_errors = Errors::default();
            outside_errors.insert_with_default_key(AppError::NotFound);
            view! { <ErrorTemplate outside_errors/> }.into_view()
        }
        trailing_slash=TrailingSlash::Exact
        >
            <main>
                <Routes>
                    // ...
                    <StaticRoute
                        mode=StaticMode::Upfront
                        path="/another-page/"
                        view=AnotherPage
                        static_params=move || Box::pin(async move { StaticParamsMap::default() })
                    />
                </Routes>
            </main>
        </Router>
```

It doesn't seem that the current static generation plays well with this, and I end up with this output:

```
building static route: /sub-page
Errors: [
    NotFound,
]
building static route: /another-page
Errors: [
    NotFound,
]
building static route:
```

which makes me think it has a hardcoded assumption on the path not ending in `/` since it also uses this to generate the filename. That of course means that the page is now just a 404 page when I visit it 😅

This seems to happen because internally the `StaticPath { path: "/sub-page/", segments: [Static("sub-page")], params: {} }` is only looking at the `segments`, which causes it to throw away the trailing `/`.

I've opened https://github.com/leptos-rs/leptos/pull/2667 with a fix that seems to work and allow me to do exactly what I was hoping for :p

**Comment by Tehnix** [comment]
@gbj would be curious if you think the fix in https://github.com/leptos-rs/leptos/pull/2667 looks good or not?

I'm also completely fine if you wanna focus on getting 0.7 out first before looking at these kinds of things :) (might also be that the PR is outdated with 0.7, in which case I can rework it after that's released)

**Comment by gbj** [comment]
Closed by #2875.
